src/helpers/mathjax.ts

```diff
@@ -0,0 +1,206 @@
+  // Install a debounce around typesetting function so that it will only run once
+  // every Xms even if called multiple times in that period
+  return debounce(typesetDocumentPromise, 100, {
+    leading: true,
+    trailing: false,
+  }).bind(null, root, windowImpl);
+});
+getTypesetDocument.cache = new WeakMap();
+
+// typesetMath is the main exported function.
+// It's called by components like HTML after they're rendered
+const typesetMath = (root: Element, windowImpl = assertWindow() as Window) => {
+  if (!document.getElementById('MathJax-Script')) {
+    startMathJax();
+  }
+
+  const hasMath = root.querySelector(COMBINED_MATH_SELECTOR);
+  // schedule a Mathjax pass if there is at least one [data-math] or <math> element present
+  if (windowImpl && windowImpl.MathJax && windowImpl.MathJax.Hub && hasMath) {
+    return getTypesetDocument(root, windowImpl)();
+  } else if (hasMath) {
+    // There are math elements present but MathJax is still loading/initializing
+    // Push to a temp queue to be processed in a StartupHook in configuredCallback
+    mathJaxLoadingQueue.push(() => getTypesetDocument(root, windowImpl)());
+  }
+
+  return Promise.resolve();
+};
+
+function startMathJax(windowImpl: Window = assertWindow() as Window): Promise<void> {
+  if (windowImpl.MathJax !== undefined || document.getElementById('MathJax-Script')) {
+    return Promise.resolve();
+  }
+
+  const script = document.createElement('script');
+  script.src = 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-MML-AM_HTMLorMML-full&delayStartupUntil=configured';
+  script.id = 'MathJax-Script';
+  script.async = true;
+  document.head.appendChild(script);
+
+  const configuredCallback = () => {
+    // there doesn't seem to be a config option for this
+    windowImpl.MathJax.HTML.Cookie.prefix = 'mathjax';
+    // proceed with mathjax initi
+    windowImpl.MathJax.Hub.Configured();
+    // Attempt another pass for calls stored during load & init. This hook is the final
+    // signal in the startup sequence, not long after MathJax.isReady becomes true:
+    // https://docs.mathjax.org/en/v2.7-latest/advanced/startup.html#startup-sequence
+    windowImpl.MathJax.Hub.Register.StartupHook('End', () => {
+      mathJaxLoadingQueue.forEach(op => op());
+      mathJaxLoadingQueue = [];
+    });
+  };
+
+  if (windowImpl.MathJax && windowImpl.MathJax.Hub) {
+    windowImpl.MathJax.Hub.Config(MATHJAX_CONFIG);
+    // Does not seem to work when passed to Config
+    windowImpl.MathJax.Hub.processSectionDelay = 0;
+    configuredCallback();
+  } else {
+    // If the MathJax.js file has not loaded yet:
+    // Call MathJax.Configured once MathJax loads and
+    // loads this config JSON since the CDN URL
+    // says to `delayStartupUntil=configured`
+    (MATHJAX_CONFIG as any).AuthorInit = configuredCallback;
+    windowImpl.MathJax = MATHJAX_CONFIG;
+  }
+
+  return Promise.resolve();
```

Comment:
same, this promise isn't really doing anything. i would prefer for it to resolve when mathjax is done loading (wrap everything in promise.new and call `resolve` from inside the callback) but if you can't or that isn't necessary or whatever then you can remove this, its not doing anything

Reply:
you could write it so typeset awaits the setup before doing the typesetting, which removes the need for the custom queue

and sometimes its helpful to know when typesetting is finished, like if you're rendering your pdf. really it'd be nice if the debounce logic included a check to make sure previous typesets were done before allowing new ones to be called, because concurrent typesets on the same element can lead to visual issues

Reply:
Agreed, I did initial try that route because it feels cleaner, but ended up having to flip things to make MathJax startup control the flow. Resolving inside the callback/startup hook works fine for the first component that calls it, but if another component calls it, I found I couldn't bail early like this:
```
return new Promise((resolve) => {
    if (windowImpl.MathJax !== undefined || document.getElementById('MathJax-Script')) {
      return resolve();
    }
```

because it could be in startup still, and not ready to typeset. Hmm what do you think about "sleeping" `startMathJax` with a setInterval until `MathJax.isReady` is true.

```
return new Promise((resolve) => {
  if (MathJax.isReady) { return resolve(); }

  const interval = setInterval(() => {
  if (MathJax.isReady) {
     resolve();
     clearInterval(interval);
   }
   }, timeout);

  ... do mathjax script injection stuff without startup signal hook
});
```

Reply:
Sorry to jump ahead but I pushed a commit with this that seems to test well, and removes the queue. I kinda like that it simplifies the reasoning and behaves like `resolveOrWait`. Thoughts?